[dynamic_layouts] Add Wrap layout and updates to the README (#2513)
flutter/packages: packages/dynamic_layouts/README.md

[developing packages and plugins](https://flutter.dev/developing-packages).
-->

TODO(DavBot02): Put a short description of the package here that helps potential users
know whether this package might be useful for them.

## Features

TODO(snat-s): List what your package can do. Maybe include images, gifs, or videos.
This package provides support for multi sized tiles and different layouts.
Currently the layouts that are implemented in this package are `Staggered` and
`Wrap`.

You can have reversed and with horizontal Grids with this layouts.

The following are some demos of how each of the grids look.

A staggered grid demo:

> <video src="assets/staggered_grid_demo.mov">

Reviewer: Some of these files are multiple MB big. That may be to large to be checked into this repository. Have you checked what other packages in this repository do to embed assets? Maybe we can move these assets to our assets repo: https://github.com/flutter/assets-for-api-docs

/cc @Piinks

<?code-excerpt "dynamic_grid_view_wrap.dart" (Example)?>
```dart
final List<Widget> children = List.generate(
   250,
   (index) => Container(
     height: index.isEven ? 100 : 50,
     width: index.isEven ? 95 : 180,
     color: index.isEven ? Colors.red : Colors.blue,
     child: Center(child: Text('Item $index')),
   ),
 );

DynamicGridView.wrap(
     mainAxisSpacing: 10,
     crossAxisSpacing: 20,
     children: children,
);
```

Here is the result of the code:

<video src="assets/simple_wrap_demo.mov">

<?code-excerpt "dynamic_grid_view_builder.dart (Example)"?>
```dart
DynamicGridView.builder(
     gridDelegate: const SliverGridDelegateWithWrapping(
        mainAxisSpacing: 20,
        childMainAxisExtent: 250,
        childCrossAxisExtent: 50,
     ),
     itemBuilder: (BuildContext context, int index) {
       return Container(
         height: 200,
         color: index.isEven ? Colors.amber : Colors.blue,
         child: Center(
           child: Text('$index'),
         ),
       );
     },
   ),
```

Here is the result of the code:

<video src="assets/wrap_demo_with_one_fixed_axis.mov">

### Staggered

> Using the Staggered layout is simple. It can be used with the

Author: The capitalization?

`DynamicGridView.stagger` and you can still use the delegates from `GridView`
like `SliverGridDelegateWithMaxCrossAxisExtent` and
`SliverGridDelegateWithFixedCrossAxisCount`.

<?code-excerpt "dynamic_grid_view_stagger.dart (Example)"?>
```dart
DynamicGridView.stagger(
      crossAxisCount: 4,
      mainAxisSpacing: 10.0,
      crossAxisSpacing: 10.0,
      children: children,
);
```

> <!-- TODO(snat-s): Add a video of DynamicGrid.stagger -->

Reviewer: Maybe place all the TODOs in comments like this so they stand out more and don't accidentally render in the real readme.

> [developing packages and plugins](https://flutter.dev/developing-packages).

Reviewer 2: Whomever lands their PR first, the other will likely have merge conflicts here. If so, let me know and I can help resolve if it is a messy merge. :)

<!-- TODO(DavBot02): Put a short description of the package here that helps potential users
know whether this package might be useful for them.-->

## Features
This package provides support for multi sized tiles and different layouts.
Currently the layouts that are implemented in this package are `Stagger` and
`Wrap`.

The following are some demos of how each of the grids look.

A stagger grid demo:

<!-- TODO(snat-s): Add stagger video demo -->

A wrap demo:

<!-- TODO(snat-s): Add wrap video demo -->

### Stagger Features

`DynamicGridView` is a subclass of `GridView` and gives access
to the `SliverGridDelegate`s that are already implemented in the Flutter
Framework. Some `SliverGridDelegate`s are `SliverGridDelegateWithMaxCrossAxisExtent` and
`SliverGridDelegateWithFixedCrossAxisCount`. This layout can be used with
`DynamicGridView.stagger`.

### Wrap Features

The Wrap layout is able to do runs of different widgets and adapt accordingly with
the sizes of the children. It can leave spacing with `mainAxisSpacing` and
`crossAxisSpacing`.

Having different sizes in only one of the axis is possible by
changing the values of `childCrossAxisExtent` and `childMainAxisExtent`. These

> values by default are ignored, but changing `childCrossAxisExtent` to be

Reviewer: The sentence about the values being ignored is confusing. What do I need to do to un-ignore the values? What extents are used while the value is ignored?

Reviewer 2: Yeah, I think users will likely better understand loose constraints versus tight constraints. By default they are loose, allowing the child to be whatever finite size it wants, if any of these extents are set by the user, they will be tight constraints for that axis, forcing that dimension of the child's size.

100 pixels, will make all of the children 100 pixels in the main axis.
This layout can be used with `DynamicGridView.wrap` and with
`DynamicGridView.builder` and `SliverGridDelegateWithWrapping` as the delegate.

## Getting started

<!-- TODO(DavBot02): List prerequisites and provide or point to information on how to start using the package. -->

## Usage

Use `DynamicGridView`s to access this layouts.
`DynamicGridView` has some constructors that have specific `SliverGridDelegate`s like
`DynamicGridView.wrap` and `DynamicGridView.stagger`. For a more efficient option

> `DynamicGridView.builder` works the same as `GridView.builder`.

Reviewer 2: This could be a bit more specific.

.wrap and .stagger use SliverChildListDelegate, and .builder uses SliverChildBuilderDelegate

### Wrap

> The following is a simple examples of how to use `DynamicGridView.wrap`.

Reviewer 2: ```suggestion
The following is a simple example of how to use `DynamicGridView.wrap`.
```

Reviewer 2: Oh actually different typo 🤣 
```suggestion
The following are a simple examples of how to use `DynamicGridView.wrap`.
```

<?code-excerpt "dynamic_grid_view_wrap.dart" (Example)?>
```dart
final List<Widget> children = List.generate(
   250,
   (index) => Container(
     height: index.isEven ? 100 : 50,
     width: index.isEven ? 95 : 180,
     color: index.isEven ? Colors.red : Colors.blue,
     child: Center(child: Text('Item $index')),
   ),
 );

DynamicGridView.wrap(
     mainAxisSpacing: 10,
     crossAxisSpacing: 20,
     children: children,
);
```

The following example uses `DynamicGridView.builder` with
`SliverGridDelegateWithWrapping`.

<?code-excerpt "dynamic_grid_view_builder.dart (Example)"?>
```dart
DynamicGridView.builder(
     gridDelegate: const SliverGridDelegateWithWrapping(
        mainAxisSpacing: 20,
        childMainAxisExtent: 250,
        childCrossAxisExtent: 50,
     ),
     itemBuilder: (BuildContext context, int index) {
       return Container(
         height: 200,
         color: index.isEven ? Colors.amber : Colors.blue,
         child: Center(
           child: Text('$index'),
         ),
       );
     },
   ),
```

> By using `childCrossAxisExtent` and `childMainAxisExtent` one of the axis

Reviewer 2: ```suggestion
By using `childCrossAxisExtent` and `childMainAxisExtent` the main axis
```

can be limited to have a specific size and the other can be set to loose
constraints.

<?code-excerpt "wrapping_fixed_axis.dart" (Example)?>
```dart
DynamicGridView.builder(
  gridDelegate: const SliverGridDelegateWithWrapping(
    mainAxisSpacing: 20,
    childMainAxisExtent: 250,
  ),
  itemBuilder: (BuildContext context, int index) {
    return Container(
      height: 200,
      color: index.isEven ? Colors.amber : Colors.blue,
      child: Center(
        child: Text('$index'),
      ),
    );
  },
),
```

### Stagger

> Using the stagger layout is simple. It can be used with the constructor

Reviewer: nit: remove the sentence about it being simple?
